# 4.6 Combining Data 2

# 4.6 Exercise Continued

## 3. In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
#set path

path = r'C:\Users\Karen\Documents\Personal\Data Analytics\Career Foundry\Achievement 4\Instacart Basket Analysis'

In [3]:
# import pickle file

df_orders_products_combined = pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data', 'orders_products_combined.pkl'))

## 4. Check the shape of the imported dataframe.

In [4]:
df_orders_products_combined.shape

(32434489, 11)

In [5]:
df_orders_products_combined

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                  0   2539329        1             1                   2   
1                  0   2539329        1             1                   2   
2                  0   2539329        1             1                   2   
3                  0   2539329        1             1                   2   
4                  0   2539329        1             1                   2   
...              ...       ...      ...           ...                 ...   
32434484     3421081   2977660   206209            13                   1   
32434485     3421081   2977660   206209            13                   1   
32434486     3421081   2977660   206209            13                   1   
32434487     3421081   2977660   206209            13                   1   
32434488     3421081   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 11 columns]

In [6]:
# Checking value count of '_merge' in df_orders_products_combined before merging with df_prods.

df_orders_products_combined['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [7]:
# Dropping the extra indexing column "Unnamed: 0"

df_orders_products_combined = df_orders_products_combined.drop(columns = ['Unnamed: 0'])

## 5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set. 

In [8]:
# Importing products data set.

df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'Prepared Data', 'products_checked.csv'), index_col=False)

In [9]:
# Checking shape of df_prods.

df_prods.shape

(49672, 6)

In [10]:
df_prods

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

In [11]:
# Dropping the extra indexing column "Unnamed: 0"

df_prods = df_prods.drop(columns = ['Unnamed: 0'])

### Considering  we want to see the product name for each product id in the df_orders_products_combined data frame, I think an inner join would be best because there would be no need to see product names for products that have not been ordered in the final merge.

In [12]:
# Merging data frames

df_orders_products_merged = df_orders_products_combined.merge(df_prods, on = 'product_id')

### I did not add the "indicator = True" argument becuase this produced an error because df_orders_products_combined already has a '_merge' column.

In [13]:
df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

In [14]:
df_orders_products_merged.shape

(32404859, 14)

## 6. Confirm the results of the merge using the merge flag.

In [15]:
df_orders_products_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

### There were 32,434,489 rows in the orders_products_combined data frame. There were 49,672 rows in the prods data frame. When merged with an inner join the result was 32,404,859 rows. This is 29,630 fewer records than in the orders_products_combined data frame.

In [16]:
# Double checking the merge with an outer join.

df_orders_products_merged_outer = df_orders_products_combined.merge(df_prods, on = 'product_id', how = 'outer')

In [17]:
df_orders_products_merged_outer['_merge'].value_counts()

_merge
both          32435059
left_only            0
right_only           0
Name: count, dtype: int64

### There are no left_only or right_only merged values

## 7. Export the newly created dataframe as orders_products_merged in a suitable format (taking into consideration the size).

In [20]:
# Exporting as pickle file

df_orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))